### ISSUE TITLE
The difference between your paper and your implementations

### ISSUE THREAD

**Issue by iRmantou** [phillipi/pix2pix#120]
Hello, @junyanz :
I have read your paper and browsed your website.I found some difference among your implementations.

**A approach:** Your Torch and Pytorch implementations and Tensorflow versions implemented by Christopher Hesse and  Yen-Chen Lin. All of them use the same architecture, for example: C64-C128-C256-C512 (n_layers = 3), which has the input of the whole images!

**B approach:** Keras implemented by Thibault de Boissiere has the same architecture as above but divide the whole images into **N patches** with the shape of 70×70 (210×210 image would be divide into 9 patches) which is your default setting in paper and the network's input is a single patch image!

So, I am confused which is your original purpose, and the most important I hope to know: different patch size Discriminators are only different in **network architectures** or in **network input shapes** either, like Keras implement

**Comment by phillipi** [comment]
Hi @iRmantou,

Both approaches are, in fact, mathematically the same! (assuming proper setting of parameters, in reality they likely differ slightly) The A approach is a fully-convolutional network, where the output neurons have receptive fields that are 70x70 patches over the input image. This is equivalent to chopping up the input into 70x70 patches explicitly, and feeding each one into a smaller network, whose output is a single scalar. Some more discussion of this equivalence is here: https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/issues/39

I like the fully-convolutional approach since it is more efficient (it shares computation between adjacent overlapping patches). I haven't looks at the Keras implementation you mention though, so there may be good reasons why they chose to explicitly chop up the image.

**Comment by iRmantou** [comment]
Thank you!!I understand!

**Comment by onursertkaya** [comment]
Hi, sorry for asking about this again after closure.

I've been trying to reimplement CycleGAN architecture without taking a look at any code (torch, PyTorch, Tensorflow etc.). What I'm gonna ask is; I couldn't find any reference for the stride of discriminator's last layer. Here, in the file `pix2pix/scripts/receptive_field_sizes.m` you have,
```
% fix the output size to 1 and derive the receptive field in the input
out = ...
f(f(f(f(f(1, 4, 1), ...   % conv4 -> conv5
             4, 1), ...   % conv3 -> conv4
             4, 2), ...   % conv2 -> conv3
             4, 2), ...   % conv1 -> conv2
             4, 2);       % input -> conv1

fprintf('n=3 discriminator receptive field size: %d\n', out);
```
It says the last **TWO** layer have stride = 1. Also in the file `pytorch-CycleGAN-and-pix2pix/models/networks.py`, starting from line 413, you have,
```
        nf_mult_prev = nf_mult
        nf_mult = min(2**n_layers, 8)
        sequence += [
            nn.Conv2d(ndf * nf_mult_prev, ndf * nf_mult,
                      kernel_size=kw, stride=1, padding=padw, bias=use_bias),
            norm_layer(ndf * nf_mult),
            nn.LeakyReLU(0.2, True)
        ]

        sequence += [nn.Conv2d(ndf * nf_mult, 1, kernel_size=kw, stride=1, padding=padw)]
```
Both of these suggest that the last convolutional layer (i.e. 256, 512) **AND** the layer after that, namely, "mapper convolutional layer with dim=1" have their strides set to 1. I couldn't find any reference to this operation in both ConditionalGAN and CycleGAN papers, other than

> After the last layer, a convolution is applied to map to a 1 dimensional output, followed by a Sigmoid function.

which only tells me that **after** the 'c512' layer, I have to add a convolutional layer to map features to 1 dimensional output. Since I haven't had a look on any code until now, it has been overwhelming for me to understand the architecture. Am I missing something in the paper ?

Thank you for your time.

**Comment by pedro-abreu** [comment]
Could you please provide a link to Thibault de Boissiere's implementation please?

**Comment by iRmantou** [comment]
@pedro-abreu you could see all types of codes in here [https://phillipi.github.io/pix2pix/]
